src/Processors/Formats/Impl/ParquetKeyValueMetadataInputFormat.cpp

```diff
@@ -0,0 +1,120 @@
+#include "ParquetKeyValueMetadataInputFormat.h"
+
+#if USE_PARQUET
+
+#    include <Columns/ColumnMap.h>
+#    include <Columns/ColumnTuple.h>
+#    include <Core/NamesAndTypes.h>
+#    include <DataTypes/DataTypeMap.h>
+#    include <DataTypes/DataTypeString.h>
+#    include <Formats/FormatFactory.h>
+#    include <Processors/Formats/Impl/Parquet/ParquetMetadataReader.h>
+#    include <arrow/api.h>
+
+namespace DB
+{
+
+static NamesAndTypesList getHeaderForParquetKeyValueMetadata()
+{
+    NamesAndTypesList names_and_types{
+        {"key_value_metadata", std::make_shared<DataTypeMap>(std::make_shared<DataTypeString>(), std::make_shared<DataTypeString>())}};
+    return names_and_types;
+}
+
+ParquetKeyValueMetadataInputFormat::ParquetKeyValueMetadataInputFormat(
+    ReadBuffer & in_, Block header_, const FormatSettings & format_settings_)
+    : IInputFormat(std::move(header_), &in_)
+    , format_settings(format_settings_)
+{
+}
```

Comment:
We should check that header contains only single column `key_value_metadata` with expected type. Because it's possible to specify the structure manually for any format:
```sql
SELECT * FROM file(data.parquet, ParquetKeyValueMetadata, 'a UInt32');
```